packages/client/test/unit/resendSubscription.test.ts

```diff
@@ -0,0 +1,290 @@
+    await waitForCondition(() => {
+        return queue.values().some((msg) => msg.content === streamMessage.getParsedContent())
+    })
+}
+
+const expectEqualMessageCollections = (actual: Iterable<Message>, expected: StreamMessage[]) => {
+    expect(Array.from(actual).map((m) => m.content)).toEqual(expected.map((m) => m.getParsedContent()))
+}
+
+describe('resend subscription', () => {
+
+    let messageFactory: MessageFactory
+
+    beforeEach(async () => {
+        const authentication = createRandomAuthentication()
+        messageFactory = new MessageFactory({
+            authentication,
+            streamId: StreamPartIDUtils.getStreamID(STREAM_PART_ID),
+            streamRegistry: createStreamRegistryCached({
+                isPublicStream: true
+            }),
+            groupKeyQueue: await createGroupKeyQueue(authentication)
+        })
+    })
+
+    const createMessages = async (count: number, type: string, msgChainId?: string): Promise<StreamMessage[]> => {
+        const result: StreamMessage[] = []
+        for (let i = 0; i < count; i++) {
+            const msgId = `${type}${(msgChainId !== undefined) ? ('-' + msgChainId) : ''}-${(i + 1)}`
+            result.push(await messageFactory.createMessage({ msgId }, { timestamp: Date.now(), msgChainId }))
+        }
+        return result
+    }
+
+    const createSubscription = (
+        resend: () => Promise<PushPipeline<StreamMessage, StreamMessage>>,
+        gapFill = true
+    ) => {
+        const eventEmitter = new EventEmitter<SubscriptionEvents>()
+        const sub = new Subscription(STREAM_PART_ID, false, eventEmitter, mockLoggerFactory())
+        initResendSubscription(
+            sub,
+            undefined as any,
+            {
+                resend
+            } as any,
+            {
+                orderMessages: true,
+                gapFill,
+                maxGapRequests: MAX_GAP_REQUESTS,
+                gapFillTimeout: 200,
+                retryResendAfterTimeout: 0
+            } as any,
+            eventEmitter,
+            mockLoggerFactory()
+        )
+        return sub
+    }
+
+    it('happy path', async () => {
+        const historicalMessages = await createMessages(3, 'historical')
+        const gapMessages = await createMessages(3, 'gap')
+        const outputMessages: Queue<Message> = new Queue()
+        const resend = jest.fn()
+            .mockImplementationOnce(() => createPushPipeline(historicalMessages))
+            .mockImplementationOnce(() => createPushPipeline(gapMessages))
+        const sub = createSubscription(resend)
+        let latestMessageWhenResendComplete: Message
+        const onResendComplete = jest.fn().mockImplementation(
+            () => latestMessageWhenResendComplete = last(outputMessages.values())!
+        )
+        sub.on('resendComplete', onResendComplete)
+        setImmediate(() => outputMessages.collect(sub))
```

Comment:
To me this reads as if "push mesages originating from outputMessages to sub" and not the other way around as it actually works.